theano/gpuarray/GpuMaxAndArgmax.py

```diff
@@ -0,0 +1,209 @@
+        if (isinstance(axis, (integer_types, numpy.integer)) or
+           (isinstance(axis, numpy.ndarray) and axis.ndim == 0)):
+            axis = [int(axis)]
+        elif isinstance(axis, (tuple, list, numpy.ndarray)):
+            axis = [int(a) for a in axis]
+        elif isinstance(axis, Variable):
+            if NoneConst.equals(axis):
+                axis = None
+            elif not isinstance(axis, TensorConstant):
+                raise TypeError("MaxAndArgmax needs a constant axis. Got %s" % axis)
+            else:
+                assert (axis.dtype.startswith("int") or axis.dtype.startswith("uint"))
+                if (isinstance(axis.data, (integer_types, numpy.integer)) or
+                   (isinstance(axis.data, numpy.ndarray) and axis.data.ndim == 0)):
+                    axis = [int(axis.data)]
+                elif isinstance(axis.data, (list, numpy.ndarray)):
+                    axis = [int(i) for i in axis.data]
+        # Make axis entries non-negative, and verify that axes are valid.
+        if isinstance(axis, list):
+            for idx in xrange(len(axis)):
+                if axis[idx] < 0:
+                    axis[idx] += X.type.ndim
+                if axis[idx] < 0 or axis[idx] >= X.type.ndim:
+                    raise ValueError('Invalid axis: %s (the number of dimensions of the '
+                                     'input is: %s)' % (axis[idx], X.type.ndim))
+        # Sort axes and make them unique.
+        axis_set = set()  # used to build "broadcastable" variable below.
+        all_axes = []
+        if isinstance(axis, list):
+            axis_set = set(axis)
+            all_axes = list(axis_set)
+            all_axes.sort()
+            if all_axes == range(X.type.ndim):
+                axis = None
+        else:
+            all_axes = range(X.ndim)
+            axis_set = set(all_axes)
+        if axis is None:
+            axis = NoneConst.clone()
+        else:
+            axis = theano.tensor.as_tensor_variable(all_axes)
+            # assert axis.ndim == 1
+        inputs = [as_gpuarray_variable(X, context_name), axis]
+        # We keep the original broadcastable flags for dimensions on which
+        # we do not perform the max / argmax.
+        broadcastable = [b for i, b in enumerate(X.type.broadcastable)
+                         if i not in axis_set]
+        outputs = [GpuArrayType(X.type.dtype, broadcastable, context_name=context_name, name='max')(),
+                   GpuArrayType(self.argmax_dtype, broadcastable, context_name=context_name, name='argmax')()]
+        return Apply(self, inputs, outputs)
+
+    def perform(self, node, inputs, outputs):
+        X, axes = inputs
+        max, max_idx = outputs
+        if axes is None:
+            axes = tuple(range(X.ndim))
+        else:
+            axes = tuple(axes)
+            # axes = tuple(int(ax) for ax in axes)
+        max[0] = theano._asarray(numpy.max(X, axes), dtype=node.outputs[0].dtype)
```

Comment:
Ok. I will try to rewrite this implementation using pygpu.

At a first glance, it seems I can use ```pygpu.reduction.reduce1``` function to compute max, but I still don't find a way to compute argmax.

I will look for that tomorrow. If I don't find a solution, I shall remove the "perform" implementation.